### ISSUE TITLE
Document several limitations

### ISSUE THREAD

**Issue by bgaillard** [whoan/docker-build-with-cache-action#48]
Hi, i'm trying to migrate one of our Docker builds to Github Actions and would like to use Docker caching features to improve our build times.

We currently have a fat Multi-stage Dockerfile with 9 stages, I encountered several limitations and think those should be documented in the project (unless somebody found workaround to prevent those limitations).

Here are what I propose to be documented.

**First limitation**

Our current setup is using Buildkit (enabled through the `DOCKER_BUILDKIT` environment variables) with experimental features (enabled by passing `--experimental` to command used to start the Docker daemon). 

When Buildkit is enabled the Docker build-with-cache action cannot work because the Docker build output is different with and without Buildkit. 

Without Buildkit the output contains `Step X/Y : FROM ...` lines which can them be parsed by the [`_get_stages()`](https://github.com/whoan/docker-build-with-cache-action/blob/master/entrypoint.sh#L62) Bash function. 

With Buildkit those lines are not their.

**Second limitation**

When we read the Docker build-with-cache action README.md we believe that caching should work in all cases without any problem.

I think its not the case because of the limits documented in this StackOverflow comment https://stackoverflow.com/questions/54574821/docker-build-not-using-cache-when-copying-gemfile-while-using-cache-from/56024061#56024061

So, the order of the `--cache-from` instruction matters but in the project the current order in use is simply `{1..$max_stage}` (see it in the [`build_image()`](https://github.com/whoan/docker-build-with-cache-action/blob/master/entrypoint.sh#L210) Bash function). With our 9 stages I tried to manually change the order but I think their are simply no solution to find ONE good order which will lead to using cache everywhere / in all stages :disappointed: . 

Do you agree with those 2 limitations and to document them inside the README.md file of the project ? 

If not do you have a solution to make caching really work with our 9 stages Multi-stage Docker build ?

**Comment by whoan** [comment]
HI @bgaillard . Thanks for letting me know of these limitations.

## First limitation

I will take a look at it as I think it can be solved and I just need to spend some time on it. Let me find some spare time to dig into it, or feel free to submit a PR.

## Second limitation

In [this branch](https://github.com/whoan/docker-build-with-cache-action/tree/name-stages), I worked in a proposal to tag the stages according to the stage names. I didn't merge that branch as I need to analyze the side effects, and the link to SO that you shared is very useful in that respect.

I think that giving names to the stages. plus adding an input like `stages_order` where you can define the desired order like `stage_1,stage_2,another_stage` would be the way to go to let you specify any particular order of the caches. Anyway, I understand you could not find a solution even ordering the stages as you wanted.

I think it's OK to document such limitation but I will spend some more time (don't have the time right now) thinking if the docker limitation can be circumvented in some way.

**Comment by bgaillard** [comment]
Hi @whoan and thanks for your reactivity on this case.

> First limitation
I will take a look at it as I think it can be solved and I just need to spend some time on it. Let me find some spare time to dig into it, or feel free to submit a PR.

With Buildkit its possible to use cache exporters, so immediately I'm not sure it would be interesting to pass to much time to support Buildkit inside this action. 

I'm thinking about adding a link to https://github.com/moby/buildkit#export-cache in the README.md with a mention to indicate to use Builkit cache exporters instead.

In a second step the action could allow to support Buildkit and internally use cache exporters instead of individual stage caches.

> I think that giving names to the stages. plus adding an input like stages_order where you can define the desired order like stage_1,stage_2,another_stage would be the way to go to let you specify any particular order of the caches. Anyway, I understand you could not find a solution even ordering the stages as you wanted.

Ok, great, it will be useful for simple Multi-stage Dockerfile where we can find an order which allow to better cache.

> I think it's OK to document such limitation but I will spend some more time (don't have the time right now) thinking if the docker limitation can be circumvented in some way.

Ok, thanks in advance for this :)

**Comment by whoan** [comment]
Hi @bgaillard. I was trying to use Buildkit built-in cache exporter with no success. My idea is to replace current functionality and use Buildkit built-in feature through a flag to activate it; something like `experimental: true`, as I (still) could not find a way to parse buildx output correctly.

As I am currently not having too much time to work on this, I added the limitations to the readme so I can close this issue. In a future release, if I happen to find the way to make it work, I will be happy to add such feature.

Thanks.

**Comment by bgaillard** [comment]
Hi @whoan and thanks for the time you took to test this feature.

You conclusions makes sense to me, Buildkit changes the output in such a way it's not possible to pick the steps in the output. 

For those who would like to use Builkit for caching it's now possible to build an image while both importing / exporting Docker cache before / after the build automatically. 

For example using it inside a Github Actions workflow could be done with something like this.

```bash
#!/bin/bash

# File '/my-project/src/.github/workflows/scripts/build.sh'

docker run \
  --rm \
  --privileged \
  -v /tmp/secrets:/tmp/secrets \
  -v /myproject/src:/tmp/src \
  -v /myproject/src:/tmp/dockerfile \
  -v $HOME/.docker:/root/.docker \
  --entrypoint buildctl-daemonless.sh \
  moby/buildkit:master \
  build \
  --frontend dockerfile.v0 \
  --local context=/tmp/src \
  --local dockerfile=/tmp/dockerfile \
  --output type=image,name=docker.registry.org/my-image:my-tag,push=true \
  --import-cache type=registry,ref=docker.registry.org/my-image:build-cache \
  --export-cache type=registry,ref=docker.registry.org/my-image:build-cache,mod=max,push=true
```

```yaml
name: My Workflow
on: [push]

jobs:
  build:
    name: Build
    runs-on: ubuntu-latest
    steps:
      ...
      - name: Build with Buildkit
        run: .github/workflows/scripts/build.sh
```

But be careful, those are very recent features, I tried to use them and it did not work while exporting images in my case (I tried both with Amazon ECR and Github Packages Docker registries).

Hope this helps.

**Comment by whoan** [comment]
No problem @bgaillard.

Today I stayed thinking of this limitation and I came up with another idea I would like to try as soon as I have time. Namely:

- If you use Buildkit, do your build as you wish, but the action should flag that Buildkit was used
- After such a build, check the flag, and if set to true, run docker build again without Buildkit activated and parse the output as usual (at this stage, hopefully the cached layers will be used)
- Then send image and stages as usual to any registry as the action does right now

In that way we don't have to wait to docker team to stabilize the builtin cache functionality in Buildkit. I don't know when I will be able to try it and add it if it works, but sharing the idea might make someone to get involved and try it.

Thanks!
